### PR TITLE
Refactor AutoScaler: metricClient, context, reconcile

### PR DESCRIPTION
**Pull request by kr11**
## Pull Request Description

Based on the suggestions in [#119 comments](https://github.com/aibrix/aibrix/issues/119#issuecomment-2423305874), we continue refactoring of the autoscaling code structure, which have done:

1. Removing the base client of Metriclient (i.e., PodMetricClient) to streamline the client implementation.
2. Eliminating the base class of AutoScaler (i.e., BaseAutoscaler).
3. Merging `reconcileKPA` and `reconcileAPA` into `reconcileCustomPA`, thereby sharing common stages and adapting different PA-specific components based on the `pa.ScalingStrategy`.

img_v3_02fs_8794b8d6-b5e2-47c1-9101-7d04ebbe7d6g

### TBD 4. `metricName` -> `metricKey`
Currently, the `metricName string` is utilized in `GetPodContainerMetric` and `GetMetricsFromPods`. Both functions also include `pods []corev1.Pod` as parameters, which already encompass the `metric_keys` information. Should we continue to use `metricName` in this context?

### TBD 5. **ScalingContext - Change from Interface to Struct**: 

I have transferred some common attributes from `KpaScalingContext` to `BaseScalingContext` to make them concise. 

I recommend retaining the `ScalingContext` interface, as the `ScalingAlgorithm.ComputeTargetReplicas`, a higher-level interface that operates above specific algorithms, requires `ScalingContext` as a parameter.

 

## Related Issues
Resolves: #[119]

Related issue list in [comments](https://github.com/aibrix/aibrix/issues/119#issuecomment-2423305874).

**Important: Before submitting, please complete the description above and review the checklist below.**

---

<details>
<summary><strong>Contribution Guidelines (Expand for Details)</strong></summary>

<p>We appreciate your contribution to aibrix! To ensure a smooth review process and maintain high code quality, please adhere to the following guidelines:</p>

<h3>Pull Request Title Format</h3>
<p>Your PR title should start with one of these prefixes to indicate the nature of the change:</p>
<ul>
    <li><code>[Bug]</code>: Corrections to existing functionality</li>
    <li><code>[CI]</code>: Changes to build process or CI pipeline</li>
    <li><code>[Docs]</code>: Updates or additions to documentation</li>
    <li><code>[API]</code>: Modifications to aibrix's API or interface</li>
    <li><code>[CLI]</code>: Changes or additions to the Command Line Interface</li>
    <li><code>[Misc]</code>: For changes not covered above (use sparingly)</li>
</ul>
<p><em>Note: For changes spanning multiple categories, use multiple prefixes in order of importance.</em></p>

<h3>Submission Checklist</h3>
<ul>
    <li>[ ] PR title includes appropriate prefix(es)</li>
    <li>[ ] Changes are clearly explained in the PR description</li>
    <li>[ ] New and existing tests pass successfully</li>
    <li>[ ] Code adheres to project style and best practices</li>
    <li>[ ] Documentation updated to reflect changes (if applicable)</li>
    <li>[ ] Thorough testing completed, no regressions introduced</li>
</ul>

<p>By submitting this PR, you confirm that you've read these guidelines and your changes align with the project's contribution standards.</p>

</details>

### CODE REVIEW

File: pkg/controller/podautoscaler/common/context.go [hunk 1]
```diff
@@ -28,10 +28,29 @@ type ScalingContext interface {
...
-	upTolerance      float64
-	downTolerance    float64
+}
+
+// NewBaseScalingContext creates a new instance of BaseScalingContext with default values.
+func NewBaseScalingContext() *BaseScalingContext {
+	return &BaseScalingContext{
+		MaxScaleUpRate:   2,     // Scale up rate of 200%, allowing rapid scaling
+		MaxScaleDownRate: 2,     // Scale down rate of 50%, for more gradual reduction
+		ScalingMetric:    "CPU", // Metric used for scaling, here set to CPU utilization
```
Comment (Jeffwan) [inline pkg/controller/podautoscaler/common/context.go:50]: This should be something has to be configured by user.

File: pkg/controller/podautoscaler/podautoscaler_controller.go [hunk 2]
```diff
@@ -89,7 +87,23 @@ func newReconciler(mgr manager.Manager) (reconcile.Reconciler, error) {
...
+	apaScaler, err := scaler.NewApaAutoscaler(
+		0,
+		// TODO: The following parameters are specific to APA.
+		//  Adjust these parameters based on your requirements for APA.
+		scaler.NewApaScalingContext(),
+	)
+	if err != nil {
+		return nil, err
+	}
+	klog.InfoS("Initialized CostumPA: APA autoscaler successfully")
```
Comment (Jeffwan) [inline pkg/controller/podautoscaler/podautoscaler_controller.go:101]: typo. -> customPA

File: pkg/controller/podautoscaler/common/context.go (hunk 1 above)
Comment (Jeffwan) [reply pkg/controller/podautoscaler/common/context.go:50]: this is OK at this moment, we should add some validation logic later

File: pkg/controller/podautoscaler/podautoscaler_controller.go (hunk 2 above)
Comment (kr11) [reply pkg/controller/podautoscaler/podautoscaler_controller.go:101]: Thanks, fix it.

File: pkg/controller/podautoscaler/common/context.go (hunk 1 above)
Comment (kr11) [reply pkg/controller/podautoscaler/common/context.go:50]: Sure, we can allow user to config it by configmap or `pa.yaml` later.

**Pull request merged**